### ISSUE TITLE
prediction differed after converting the model

### ISSUE THREAD

**Issue by jayle19930918** [tobybreckon/fire-detection-cnn#31]
Hi, professor! After commented out the dropout layer, I got the tensorflow pb file and the opencv dnn test worked well. Then, I converted the .pb file to mxnet models with MMdnn successfully. But the predictions is inconsistent  the opencv dnn test. Before prediction, the image the resized to 224*224. I want to know whether data preprocessing is required. Thank you!

**Comment by tobybreckon** [comment]
Please post the prediction vector output for the first 10 frames of the test video - for both the OpenCV DNN test, and your mxnet test side-by-side. e.g.

```
OpenCV DNN             Mxnet
[Val, Val].                    [Val, Val]
```

or similar.

Each video frame or image should be resized to 224*224 before being passed into the network.

As per earlier issues, the effect of removing the dropout at inference runtime is unknown and may be framework dependant - although it is used during training, in theory all the weights need to be halfed/downscaled by the dropout rate. Where does this happen when you switch framework ? (I am unsure - I can imagine the weights being saved pre-scaled or otherwise and being scaled at inference in the fly). If you remove the dropout, what happens? (tbc)

**Comment by jayle19930918** [comment]
> Please post the prediction vector output for the first 10 frames of the test video - for both the OpenCV DNN test, and your mxnet test side-by-side. e.g.
> 
> ```
> OpenCV DNN             Mxnet
> [Val, Val].            [Val, Val]
> ```
> 
> or similar.
> 
> Each video frame or image should be resized to 224*224 before being passed into the network.
> 
> As per earlier issues, the effect of removing the dropout at inference runtime is unknown and may be framework dependant - although it is used during training, in theory all the weights need to be halfed/downscaled by the dropout rate. Where does this happen when you switch framework ? (I am unsure - I can imagine the weights being saved pre-scaled or otherwise and being scaled at inference in the fly). If you remove the dropout, what happens? (tbc)

Now, the predictions of converted model are [nan, nan] with python 3.6.9. I printed the layer name of firenet.pb and the converted model.  The converted model did not have MaxPooling layer and LRN layer compared to the original model. The original model worked well in the opencv dnn test. I think the problem arised from model conversion.

**Comment by tobybreckon** [comment]
This seems a pretty good conclusion.

AFIAK, other people have successfully converted these models as inputs to MMdnn but perhaps not as output to Mxnet where the problem may be.

Option 1: reproduce architecture definition in MXnet, work out way to manually copy weights or use another tool to read in .pb file

Option 2: reproduce architecture definition in MXnet and retrain from provided dataset

Option 3: try a different DL framework or conversion tools ..

**Comment by tobybreckon** [comment]
Are you using ``mmconvert`` in MMdnn?

If so can you send me the full command you are using?

**Comment by jayle19930918** [comment]
> Are you using `mmconvert` in MMdnn?
> 
> If so can you send me the full command you are using?

Yes, I used mmconvert in MMdnn. The flowing is the full command:

mmconvert -sf tensorflow -iw firenet.pb --inNodeName InputData/X --inputShape 224,224,3 --dstNodeName FullyConnected_2/Softmax -df mxnet -om mxnet_firenet -dw mxnet_firenet-0000.params

**Comment by Luyubo** [comment]
> Hi, professor! After commented out the dropout layer, I got the tensorflow pb file and the opencv dnn test worked well. Then, I converted the .pb file to mxnet models with MMdnn successfully. But the predictions is inconsistent the opencv dnn test. Before prediction, the image the resized to 224*224. I want to know whether data preprocessing is required. Thank you!



> Hi, professor! After commented out the dropout layer, I got the tensorflow pb file and the opencv dnn test worked well. Then, I converted the .pb file to mxnet models with MMdnn successfully. But the predictions is inconsistent the opencv dnn test. Before prediction, the image the resized to 224*224. I want to know whether data preprocessing is required. Thank you!

Hi, I want to use the opennnv dnn module to call firenet.pb, but after the conversion, I found that only .pb files, no .pbtxt files, what should I do? I am a newcomer to deep learning. I want to know how to code to implement opencv to call the firenet model.

**Comment by tobybreckon** [comment]
@jayle19930918 - thanks I will take a look directly when I get a moment. Can I confirm you are using the latest version of all the tools? (otherwise we may be chasing a bug that is fixed)

@Luyubo - for a new issue, create a new thread and also use the provided template, please. The provided opencv test example in the convertor directory shows exactly how to use the DNN module with only a on file.

**Comment by jayle19930918** [comment]
> @jayle19930918 - thanks I will take a look directly when I get a moment. Can I confirm you are using the latest version of all the tools? (otherwise we may be chasing a bug that is fixed)
> 
> @Luyubo - for a new issue, create a new thread and also use the provided template, please. The provided opencv test example in the convertor directory shows exactly how to use the DNN module with only a pb file.

@tobybreckon -Thank you! I have tried both stable version and the newest version, the results were same. Before convertion, the input shape is 1*224*224*3. After conversion, the input shape changes to 1*3*224*224 and we need to replace 'Placeholder' with 'data' in .json file. Otherwise, the inference will fail.

**Comment by tobybreckon** [comment]
@jayle19930918 - to confirm is this a solution to your problem ?

i.e. to use with MxNet you have to ...

- commented out the dropout layer in ```firenet.py``
- run conversion to .pb file
- Run:  ```mmconvert -sf tensorflow -iw firenet.pb --inNodeName InputData/X --inputShape 224,224,3 --dstNodeName FullyConnected_2/Softmax -df mxnet -om mxnet_firenet -dw mxnet_firenet-0000.params```
- edit .json output of this command (which file?) to replace string 'Placeholder' with 'data' 
- then use with MxNet but input the image as size 1x3x224x224, not 1x224x224x3

and then it works with MxNet ? (is the side-by-side output the same ?)

**Comment by tobybreckon** [comment]
This is beyond my immediate knowledge base, sorry.

I don't use MXnet but is the image is being resized (1x3x244x244) fed in correctly (as float ?), etc ?
To my mind the 1x3x224x224 bit seems wrong.

**Comment by jayle19930918** [comment]
> This is beyond my immediate knowledge base, sorry.
> 
> I don't use MXnet but is the image is being resized (1x3x244x244) fed in correctly (as float ?), etc ?
> To my mind the 1x3x224x224 bit seems wrong.

I also think the input shape being resized to 1x3x244x244 is a bit strange. I plan to retrain from provided dataset using MXNet. Thank you for analyzing my problem!
